Marlin/src/HAL/STM32/tft/xpt2046.cpp

```diff
@@ -0,0 +1,185 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2019 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "../../../inc/MarlinConfig.h"
+
+#if ENABLED(TOUCH_SCREEN)
```

Comment:
Is this XPT2046 needed for all `TOUCH_SCREEN` or just those that enable `TOUCH_BUTTONS` (now called `HAS_XPT2046_BUTTONS`) ?

Reply:
All touch screens use it. It can be HW or SW SPI. 

The current marlin xpt code is: `Marlin/src/feature/touch/xpt2046.cpp` (SW SPI). It is used only by Emulated DOGM FSMC (FSMC_GRAPHICAL_TFT).

MKS implemented their own xpt (HW SPI) in `Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp`.

This PR will add one more xpt in `Marlin/src/HAL/STM32/tft/xpt2046.cpp` (seems it supports both SW and HW SPI). 

The first two use libmaple (I'm not sure if Marlin have any SPI abstraction on top it).
This last one use direct Arduino_Core_STM32 API.

I'm aware of this, because I'm creating support for Emulated DOGM using SPI (to support the marlin text menus in the new MKS boards that are SPI only)

Reply:
With LibMaple going the way of the dodo, I guess we should focus on unifying all the XPT2046 support under a single API built around Arduino core.

Reply:
There are two drivers for XPT2046 now.
I am not counting the one in mks_ui as in its current state it is hard-coded for specific MKS board.


The old one in `feature/touch/xpt2046.cpp` is used with mapple-based HAL STM32F1 and upscaled 128x64 interface.
It is enabled by HAS_XPT2046_BUTTONS (ex. TOUCH_BUTTONS). This driver handles both IO and UI-related logic.

The new one in `HAL/STM32/tft/xpt2046.cpp` is used with new MarlinUI implementations for TFT screens. 
It is enabled by TOUCH_SCREEN. New driver is ding IO only. All UI-related logic is handled by platform-independent code in `lcd/tft/touch.cpp`
For the sake of compatibility the pins configuration is the same for both implementation and new implementation is backward compatible with HAS_XPT2046_BUTTONS calibration data (in landscape orientation).

Both implementations can use software and hardware SPI and both implementations can use optional touch irq pin.
New implementation is DMA-aware, to support background data transfer to TFT screen via shared SPI (shared between TFT and touch, SD card should use another SPI of SDIO)

From where I stand the dogm emulation is a 'legacy' solution which is to be replaced by new UI implementation (along with migration to new HAL).
Upscaled 128x64 interface with touch buttons is still available in old HAL STM32F1, but I have no plans to implement is for new STM32 HAL. Same for old XPT2046 driver.

Reply:
This particular XPT2046 is needed for all `TOUCH_SCREEN`

Reply:
Okay, makes sense. So we have a touch-screen-with-xpt and then we have some screens that could have xpt touch too.